### ISSUE TITLE
Fuzzing cwltool

### ISSUE THREAD

**Issue by mr-c** [common-workflow-language/cwltool#1170]
https://alexgaynor.net/2015/apr/13/introduction-to-fuzzing-in-python-with-afl/

> [Fuzzing](https://en.wikipedia.org/wiki/Fuzz_testing) is a technique in computer testing and security where you generate a bunch of random inputs, and see how some program handles it. For example, if you had a JPEG parser, you might create a bunch of valid images and broken images, and make sure it either parses them or errors out cleanly. In C (and other memory unsafe languages) fuzzing can often be used to discover segfaults, invalid reads, and other potential security issues. Fuzzing is also useful in Python, where it can discover uncaught exceptions, and other API contract violations.

Goal: No interaction with `cwltool` should run forever, produce a segfault, or quit with just a plain Python exception traceback. 

- [ ] do once manually
- [ ] fix bugs, add test cases
- [ ] if many bugs are found, research [test case reduction](http://embed.cs.utah.edu/creduce/)
- [ ] later, automate; maybe run on ci.commonwl.org when there are no other jobs?

**Comment by kunal12298** [comment]
Hey, looking forward to contribute for this project for GSOC, can you let me know how to get started?

**Comment by mr-c** [comment]
Welcome @kunal12298 ! Do you have any experience with fuzzing and/or Python? Did you read the blog post link above?

**Comment by kritika12298** [comment]
No, I don't have experience with Fuzzing but I know Python. Yes, I have read the link above and it would be helpful if you guide me on how to get started. I will work hard and prove myself.

**Comment by kritika12298** [comment]
Yes, I have read the blog post link above and though I have an experience with Python but I have no experience with Fuzzing. Want to complete this project as soon as possible in order to submit my proposal early, so it would be helpful if you will guide me.

**Comment by kunal12298** [comment]
Thank you and yes, I have read the blog post link above.
I know Python but have no experience with Fuzzing.
If guided, then I will surely prove myself

**Comment by mr-c** [comment]
@kunal12298 I recommend working through the CWL User Guide so that you get experience with how to use `cwltool` https://www.commonwl.org/user_guide/ Then, based upon that experience, we should be able to work together to create a plan for fuzzing `cwltool`. We are also new to fuzzing! :-)

**Comment by Shubhupatel** [comment]
Hello sir, I am interested in contributing to this project for gsoc. And I am currently read about this project. I would appreciate some help for a better and quick understanding.
Thank you!

**Comment by mr-c** [comment]
Hello @Shubhupatel, thanks for your interest. As I suggested to @kunal12298 , please work through the CWL user guide and the other linked resources. I've also updated the first post above as well.

**Comment by Shubhupatel** [comment]
Sir can I know about project proposal format. like you can suggest me any specific proposal format or I will make my own.

**Comment by chhavi18387** [comment]
Hello Sir, I am really interested in contributing to this project. Please guide me further.

**Comment by chhavi18387** [comment]
Hey!
I am really interested in contributing to this project as a part of GSoC'20. Please guide me further.

**Comment by mr-c** [comment]
@Shubhupatel Please follow the advice at https://www.open-bio.org/events/gsoc/gsoc-project-ideas/#fuzz-cwl under "how to apply" and https://developers.google.com/open-source/gsoc/help/student-advice

**Comment by dark00infinity** [comment]
Hi,
My name is Deepak singh and I am a second year Electronics and Communication Engineering student at university institute of engineering and technology, Panjab University. I can code in Python ,c++and c. I take special interest when it comes to exploring and executing new algorithms and modules. I am a resident of India. Yes, I have knowledge about fuzzing. I have also made a generation based python fuzzer in one of my hackathon and also have experience with PyFuzz module in python. i am new to cwl but started learning it. I want to participate in GSOC 2020.Therefore want some more insights on what kind of fuzzer is the requirement here(any speifications) and some more elaboration about this repository [CWL 1.1 conformance tests](https://github.com/common-workflow-language/cwl-v1.1/tree/master/tests)

**Comment by Shubhupatel** [comment]
> Hello @Shubhupatel, thanks for your interest. As I suggested to @kunal12298 , please work through the CWL user guide and the other linked resources. I've also updated the first post above as well.

Sir what we do next after reading and some work on CWL user guide.

**Comment by utkarsh1148** [comment]
Hi, My name is Utkarsh Mishra Second Year student at Thapar Institute Patiala India. I have worked with CWL and fuzzing before and am excited to contribute to this project. I liked the project idea and started to work on this. I wanted to ask if there is any slack or Discord group for this or should i post my progress here.

**Comment by muhammadusman93** [comment]
Hi Michael,

I am Usman and a third year PhD student at the University of Texas at Austin, USA. I am doing research in automated software testing and want to contribute to the Fuzzing Project for CWL. 

I have drafted a proposal and wanted to share it with you for feedback. Kindly guide me about how I can discuss it with you.

Thanks,

M. Usman

**Comment by mr-c** [comment]
Hello all. Please share your draft proposals with us via mrc@commonwl.org & peter.amstutz@curii.com
